flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/DataStreamCsvITCase.java

```diff
@@ -82,41 +102,86 @@ public void testBoundedTextFileSourceWithJackson() throws Exception {
-                    new CitiesPojo(
-                            "San Francisco",
-                            new BigDecimal("37.7562"),
-                            new BigDecimal("-122.4430"),
-                            "United States",
-                            "US",
-                            "California",
-                            "",
-                            3592294L),
-                    new CitiesPojo(
-                            "Beijing",
-                            new BigDecimal("39.9050"),
-                            new BigDecimal("116.3914"),
-                            "China",
-                            "CN",
-                            "Beijing",
-                            "primary",
-                            19433000L)
-                };
-
-        assertEquals(Arrays.asList(expected), result);
+        final CsvReaderFormat<CityPojo> csvFormat = CsvReaderFormat.forPojo(CityPojo.class);
+        final List<CityPojo> result = initializeSourceAndReadData(testDir, csvFormat);
+
+        assertEquals(Arrays.asList(pojos), result);
+    }
+
+    @Test
+    public void testCustomBulkWriter() throws Exception {
+        final File outDir = TMP_FOLDER.newFolder();
+
+        StreamingFileSink<CityPojo> sink =
+                StreamingFileSink.forBulkFormat(
+                                new Path(outDir.toURI()), new TestCsvBulkWriterFactory())
+                        .withBucketAssigner(new BasePathBucketAssigner<>())
+                        .withBucketCheckInterval(10L)
+                        .withRollingPolicy(build())
+                        .build();
+
+        OneInputStreamOperatorTestHarness<CityPojo, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(new StreamSink<>(sink), 1, 1, 0);
+
+        testHarness.setup();
+        testHarness.open();
+
+        testHarness.processElement(new StreamRecord<>(pojos[0], 1L));
+        testHarness.processElement(new StreamRecord<>(pojos[1], 1L));
+        testHarness.processElement(new StreamRecord<>(pojos[2], 1L));
+        testHarness.snapshot(1L, 1L);
+        testHarness.notifyOfCompletedCheckpoint(2L);
+
+        Map<File, String> contents = getFileContentByPath(outDir);
+        Map.Entry<File, String> onlyEntry = contents.entrySet().iterator().next();
+        String[] result = onlyEntry.getValue().split("\n");
+
+        assertThat(result, is(CSV_LINES));
+        //        assertThat(result, arrayContaining(CSV_LINES));
+    }
+
+    static Map<File, String> getFileContentByPath(File directory) throws IOException {
+        Map<File, String> contents = new HashMap<>(4);
+
+        final Collection<File> filesInBucket = FileUtils.listFiles(directory, null, true);
+        for (File file : filesInBucket) {
+            contents.put(file, FileUtils.readFileToString(file));
+        }
+        return contents;
+    }
+
+    static final class TestCsvBulkWriterFactory implements BulkWriter.Factory<CityPojo> {
```

Comment:
This feels a bit awkward, how about

```
class CsvBulkWriter {
  static <T> BulkWriter.Factory<T> forPojo(Class<T> pojoClass) {
  
            final ObjectMapper objectMapper = new ObjectMapper();
            PojoConverter.ConverterContext converterContext =
                    new PojoConverter.ConverterContext(objectMapper);
            final PojoConverter converter = createConverter(); // see next comment
            
            final CsvMapper csvMapper = new CsvMapper();
            final CsvSchema schema = csvMapper.schemaFor(pojoClass);
    new CsvBulkWriter.Factory(schema, csvMapper, converter); // basically same ctor as CsvFormat with converter going into the opposite direction
  }
}

Reply:
Added native POJOs support in CsvBulkWriter.